src/ex_cmds2.c

```diff
@@ -181,10 +181,30 @@ dialog_changed(
 	// May get file name, when there is none
 	browse_save_fname(buf);
 #endif
-	if (buf->b_fname != NULL && check_overwrite(&ea, buf,
-				    buf->b_fname, buf->b_ffname, FALSE) == OK)
+	int empty_buf = buf->b_fname == NULL ? TRUE : FALSE;
+	if (empty_buf == TRUE)
+	{
+	    buf->b_fname = vim_strsave((char_u *)"Untitled");
+	    if (STRLEN(NameBuff) == 0 && mch_dirname(NameBuff, MAXPATHL) == OK)
+	    {
+		buf->b_ffname = vim_strsave(NameBuff);
+	    }
+	}
+	if (check_overwrite(&ea, buf, buf->b_fname, buf->b_ffname, FALSE) == OK)
+	{
 	    // didn't hit Cancel
-	    (void)buf_write_all(buf, FALSE);
+	    if (buf_write_all(buf, FALSE) == OK)
+		return;
+	}
+
+	// resotre to empty when write failed
+	if (empty_buf)
```

Comment:
Style: here you are using `if(empty_buf)`, above you were using `if(empty_buf == TRUE)`